signal.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2022 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fx
+
+import (
+	"fmt"
+	"os"
+	"os/signal"
+	"sync"
+)
+
+// Signal represents a operating system process signal.
+type Signal struct {
+	OS os.Signal
+}
+
+// String will render a Signal type as a string suitable for printing.
+func (sig Signal) String() string {
+	return fmt.Sprintf("%v", sig.OS)
+}
+
+type signalReceivers struct {
+	last     *Signal
+	lastLock sync.RWMutex
+	dones    []chan os.Signal
+	doneLock sync.RWMutex
+}
+
+func (recv *signalReceivers) done() chan os.Signal {
```

Comment:
minor habitual thing:
for private/unexported types like signalReceivers,
consider still exporting its methods (as long as the type is private)
to differentiate between the type's public vs private API.

Like, with signalReceivers, 'done' and 'broadcast' are meant to be called by app.go, 'broadcastSignal' is not, but someone reading this code wouldn't know that.
If done and broadcast were exported, that would be clearer.

Reply:
I had a side discussion with @sywhang about this the other day, and private seemed like the way to go, but I see your point about it helping a reader understanding that "although this type is private, it's API is these methods", when using public methods. I'll update it.

Reply:
This should be fixed now.